src/platforms/atomic-kms/server/kms/display_buffer.cpp

```diff
@@ -335,6 +339,78 @@ void mga::DisplaySink::set_next_image(std::unique_ptr<Framebuffer> content)
     }
 }
 
+auto mga::DmaBufDisplayAllocator::framebuffer_for(std::shared_ptr<DMABufBuffer> buffer) -> std::unique_ptr<Framebuffer>
+{
+    auto plane_descriptors = buffer->planes();
+
+    assert(plane_descriptors.size() <= 4);
+
+    auto width = buffer->size().width;
+    auto height = buffer->size().height;
+    auto pixel_format = buffer->format();
+
+    uint32_t bo_handles[4] = {0};
+    uint32_t pitches[4] = {0};
+    uint32_t offsets[4] = {0};
+
+    for (std::size_t i = 0; i < std::min(4zu, plane_descriptors.size()); i++)
+    {
+        bo_handles[i] = plane_descriptors[i].dma_buf;
```

Comment:
This is where the problem lies; `drmModeAddFB2` expects GEM handles, and you're giving it dmabuf file descriptors.

[Here](https://gitlab.freedesktop.org/mesa/drm/-/blob/main/xf86drm.h?ref_type=heads#L816) is the necessary DRM API for converting the two, along with an ominous warning about the sharp edges of that API.

Since I think we share the underlying FD with the `RenderingPlatform` (and so with EGL etc), I think we'll need to do the “everything goes through gbm” route. *That* means going through [`gbm_bo_import`](https://gitlab.freedesktop.org/mesa/mesa/-/blob/a33bd78a543ca6de3b54fa78f2eb4f2403553876/src/gbm/main/gbm.h#L406) (with `type = GBM_BO_IMPORT_FD_MODIFIER` and `flags = GBM_BO_USE_SCANOUT`) and then pulling all the GEM handles out with [`gbm_bo_get_handle_for_plane`](https://gitlab.freedesktop.org/mesa/mesa/-/blob/a33bd78a543ca6de3b54fa78f2eb4f2403553876/src/gbm/main/gbm.h#L481) (and then handling the lifetime of the `gbm_bo`, keeping a track of it, and destroying it when necessary)